crates/apollo-compiler/src/database/hir_db.rs

```diff
@@ -253,201 +201,80 @@ fn schema(db: &dyn HirDatabase) -> Arc<SchemaDefinition> {
     Arc::new(schema_def)
 }
 
-fn schema_extensions(db: &dyn HirDatabase) -> Arc<Vec<SchemaExtension>> {
-    let schema_ext = all_type_definitions(db)
-        .filter_map(|(id, definition)| match definition {
-            ast::Definition::SchemaExtension(def) => Some(SchemaExtension {
-                directives: directives(def.directives(), id),
-                root_operation_type_definition: root_operation_type_definition(
-                    def.root_operation_type_definitions(),
-                    id,
-                ),
-                loc: location(id, def.syntax()),
-            }),
-            _ => None,
-        })
-        .collect();
-    Arc::new(schema_ext)
-}
-
-fn object_types(db: &dyn HirDatabase) -> Arc<Vec<ObjectTypeDefinition>> {
-    let objects = all_type_definitions(db)
-        .filter_map(|(id, definition)| match definition {
-            ast::Definition::ObjectTypeDefinition(obj_def) => object_type_definition(obj_def, id),
-            _ => None,
-        })
-        .collect();
-    Arc::new(objects)
-}
-
-fn object_type_extensions(db: &dyn HirDatabase) -> Arc<Vec<ObjectTypeExtension>> {
-    let objects = all_type_definitions(db)
-        .filter_map(|(id, definition)| match definition {
-            ast::Definition::ObjectTypeExtension(def) => Some(ObjectTypeExtension {
-                directives: directives(def.directives(), id),
-                name: name(def.name(), id)?,
-                implements_interfaces: implements_interfaces(def.implements_interfaces(), id),
-                fields_definition: fields_definition(def.fields_definition(), id),
-                loc: location(id, def.syntax()),
-            }),
-            _ => None,
-        })
-        .collect();
-    Arc::new(objects)
-}
-
-fn scalars(db: &dyn HirDatabase) -> Arc<Vec<ScalarTypeDefinition>> {
-    let scalars = all_type_definitions(db)
-        .filter_map(|(id, definition)| match definition {
-            ast::Definition::ScalarTypeDefinition(scalar_def) => scalar_definition(scalar_def, id),
-            _ => None,
-        })
-        .collect();
-    let scalars = built_in_scalars(scalars);
-
-    Arc::new(scalars)
+macro_rules! by_name {
+    ($db: ident, $convert: expr) => {{
```

Comment:
It’s a macro because it calls `.name()` inherent methods of different types. The most straightforward way to make this not a macro would be to add a trait for that method, but that seems heavyweight since there aren’t other uses.

However, thinking about it again now, we could have the `convert` callback return `Option<(String, HirType)>` with the name instead of `Option<HirType>`. I’ll give it a go

Reply:
Ah but the next macro also pushes to `extensions` fields of different structs, and I can’t seem to find a way to express that with generics that feels nicer than macros.

Reply:
Gotcha, yeah I guess we'd have to have a trait and implement it for each relevent struct, which would be easiest to do with a macro anyways 😝 this is the better solution!